chunky/src/java/se/llbit/math/Octree.java

```diff
@@ -364,66 +374,53 @@ public int getDepth() {
   }
 
   /**
-   * Replace the implementation for the packed one
+   * Switch between any two implementation by reusing the load and store methods of
+   * the octree implementations
+   * @param newImplementation The new Octree implementation
    */
-  public void pack() {
-    if(usePacked) {
-      if(implementation instanceof NodeBasedOctree) {
-        try {
-          implementation = new BigPackedOctree(implementation.getDepth(), ((NodeBasedOctree) implementation).root);
-        } catch(PackedOctree.OctreeTooBigException e) {
-          // If octree is too big, do nothing, keep the node based implementation
-        }
-      }
-    }
-  }
-
-  private void switchToNodeBased() {
-    if(implementation instanceof PackedOctree) {
-      implementation = ((PackedOctree) implementation).toNodeBasedOctree();
+  public void switchImplementation(String newImplementation) throws IOException {
+    ImplementationFactory factory = getImplementationFactory(newImplementation);
+    if(factory.isOfType(implementation)) {
+      // Already correct implementation
+      return;
     }
-  }
-
-
-  static {
-    factories.put("NODE", new ImplementationFactory() {
-      @Override
-      public OctreeImplementation create(int depth) {
-        return new NodeBasedOctree(depth, new Node(0));
-      }
-
-      @Override
-      public OctreeImplementation load(DataInputStream in) throws IOException {
-        return NodeBasedOctree.load(in);
-      }
-    });
-
-    factories.put("PACKED", new ImplementationFactory() {
-      @Override
-      public OctreeImplementation create(int depth) {
-        return new PackedOctree(depth);
-      }
 
-      @Override
-      public OctreeImplementation load(DataInputStream in) throws IOException {
-        return PackedOctree.load(in);
+    // Here we use a PipedStream to pipe what the old implementation stores
+    // to what the new will load.
+    // With this method both octree will exist side-by-side during the conversion
+    // If this proves to be a problem, we could store to a temporary file instead
+    long nodeCount = implementation.nodeCount();
+    AtomicReference<OctreeImplementation> newOctree = new AtomicReference<>();
+    PipedOutputStream out = new PipedOutputStream();
+    PipedInputStream in = new PipedInputStream(out);
+    Thread readerThread = new Thread(() -> {
+      try {
+        OctreeImplementation impl = factory.loadWithNodeCount(nodeCount, new DataInputStream(in));
+        newOctree.set(impl);
+      } catch(IOException e) {
+        e.printStackTrace();
       }
```

Comment:
I have a lot of try/catch around here for exceptions that i don't want to leak (well i don't want to have to annotate every method with a `throws` clause but i would be ok rethrowing them as runtime exception). This is not very nice, what should i do?

Reply:
Why does this method spawn a new thread and then join it? It could just do the work in the current thread instead and the caller could decide which thread to run this in. That way, you could just throw the IOException (I think it's fine to allow throwing IOExceptions if octree conversion fails).

Reply:
The thread is here so i can use PipedStream, this kind of stream block when the buffer is full/empty and so the consumer and the producer must be on different threads.
If instead i go with the save in temporary file and reload, i wouldn't need a thread

Reply:
I changed this function to use a temporary file on disk as we discussed